mmtk/src/lib.rs

```diff
@@ -77,6 +77,7 @@ pub struct OpenJDK_Upcalls {
     pub number_of_mutators: extern "C" fn() -> usize,
     pub schedule_finalizer: extern "C" fn(),
     pub prepare_for_roots_re_scanning: extern "C" fn(),
+    pub enqueue_references: extern "C" fn(objects: &[ObjectReference], len: usize),
```

Comment:
Slice type `[T]` is not FFI-safe in Rust.  Rust will warn you if you define a function like this:
```rust
extern "C" fn enqueue_references(objects: &[ObjectReference], len: usize) { // WARNING: `extern` fn uses type `[ObjectReference]`, which is not FFI-safe
   ...
}
```

But somehow Rust doesn't give a warning if we define a function pointer (`extern "C" fn(&[T])`) with `extern "C"`. It should give a warning.

From the disassembly, I see that a `&[T]` argument is actually passed via two registers.  On x86_64, `RDI` holds the pointer to the underlying bufer, and `RSI` holds the length of it.  Your code **happens to work as expected** because the actual function 
```C
void mmtk_enqueue_references(void** objects, size_t len)
```
takes a `size_t len` as its second argument, which happens to use the register (`RSI`) for the length of the `&[T]` slice passed from Rust. If you add a third parameter `foo` to it:
```c
void mmtk_enqueue_references(void** objects, size_t len, size_t foo)
```
The third parameter `foo` will receive the actual `len: usize` argument passed from Rust (via register `RDX`).

To properly pass a slice `let s: &[T] = ...` to C, use `s.as_ptr()` to get the pointer, and use `s.len()` to get its length.

Reply:
Thanks for pointing this out. Fixed.